### PR TITLE
nixos: fix kernel override logic

### PR DESCRIPTION
**Pull request by ghost**
###### Motivation for this change

Unfortunately, after [patch](https://github.com/NixOS/nixpkgs/commit/da36847d925058fd86f027b64cc712c57be11ad8) for pull request #15095 was applied, kernel derivation build parameters override logic was broken. For example, [this](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/testing/minimal-kernel.nix#L16) code will not work anymore. So, neither `configurePhase`, nor `postUnpack`, nor any other parameter can be overriden or set anymore. This disables any changes to kernel build logic from user-side without patching the `nixpkgs` sources.
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [x] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

This patch will provide new logic for kernel build parameter override process. Also it fixes the `minimal-kernel.nix` to fit the new logic.

---

/cc @cstrahan

### CODE REVIEW

File: nixos/modules/testing/minimal-kernel.nix [hunk 1]
```diff
@@ -1,28 +1,42 @@
...
+
+  uniqueAttrs = list: key:
+    if list == [] then []
+    else
+      let
+        x = head list;
+        xs = uniqueAttrs (lib.drop 1 list) key;
+      in [x] ++ removeAttrs x key xs;
+
+  uniqueReqKernConfig = uniqueAttrs config.system.requiredKernelConfig "configLine";
```
Comment (tk-ecotelecom) [inline nixos/modules/testing/minimal-kernel.nix]: Unfortunately, the `unique` must be used, or the test will fail with an error.
I changed the code, please review.

**Comment by tk-ecotelecom** [comment]
@cstrahan, please comment on this.

**Comment by cstrahan** [comment]
@tk-ecotelecom Sorry, just now seeing this -- thank you for the second ping. I'm looking over this now.

**Comment by cstrahan** [comment]
So, I'm still poking around trying to find what could be broken (you haven't mentioned *why* anything would break). I'm a bit confused though: if I'm not mistaken, that code is hit when running the tests on Hydra in order to build the test VMs, but it looks like tests are still operating just fine: https://hydra.nixos.org/job/nixos/trunk-combined/tested#tabs-constituents

**Comment by cstrahan** [comment]
Oh, I guess that does depend on whether one has passed `minimal = true`...

**Comment by joachifm** [comment]
@cstrahan in case you didn't see https://github.com/NixOS/nixpkgs/issues/19698, we also had problems with grsecurity

**Comment by cstrahan** [comment]
@joachifm @tk-ecotelecom Ah, at least part of this makes perfect sense to me. When using `.override` on a "vanilla" kernel, that clobbers any previously specified `kernelPatches`, which is why I do this in the `apply`:

```
     apply = kernelPackages: kernelPackages.extend (self: super: {
        kernel = super.kernel.override {
          kernelPatches = super.kernel.kernelPatches ++ kernelPatches; # <======
        };
      });
```

If I don't do that, the user will lose their `kernelPatches` specified in e.g. `packageOverrides` (the same thing applies to any kernels specified in `all-packages.nix` that use the `.override` mechanism to add packages).

The case where it breaks is with kernel packages like the grsecurity one because they unconditionally add their corresponding `kernelPatches`, so the aforementioned NixOS option ends up duplicating them.

One solution, and the one I think we ought to go with, would be to make the "oddball" kernels follow the example of the "vanilla" ones -- that is, `.override { kernelPatches = ...; }` should clobber the patches, and the user would thread through the original package's patches if they wanted additional patches (e.g. `grsec.override { kernelPatches = super.grsec.kernelPatches ++ [ ... ]; }`).

How does that sound?

**Comment by cstrahan** [comment]
@tk-ecotelecom I believe the solution proposed in my last comment should fix everything, but if I'm missing some other dimension to the problem, please do let me know.

As for `boot.kernelOverrides`, I'm a little reticent to add such an option. If I understand correctly, well behaved kernel packages (and I would argue that `minimal-kernel.nix` is *not* an example of a well behaved package) should work in the presence of any overrides given in `packageOverrides`, or by setting a custom kernel via `boot.kernelPackages`. It's a little subjective, but I figured kernel patches were a common enough occurrence that they demanded their own option (I've corresponded with a number of users who wanted to quickly test out some patches but were blocked by not knowing how and where to override the package).

**Comment by tk-ecotelecom** [comment]
Well, I think, one should fix `minimal-kernel.nix` this way:
```nix
{ config, pkgs, lib, ... }:

with builtins;

let
  configfile = toFile "config" ((lib.concatStringsSep "\n"
    (lib.unique (catAttrs "configLine" config.system.requiredKernelConfig))
  ) + "\n");

  kernel = lib.makeOverridable (kernel: lib.overrideDerivation kernel (origAttrs: {
    allowImportFromDerivation = true;

    configurePhase = ''
      runHook preConfigure
      mkdir -p ../build
      make $makeFlags "''${makeFlagsArray[@]}" mrproper
      make $makeFlags "''${makeFlagsArray[@]}" KCONFIG_ALLCONFIG=${configfile} allnoconfig
      runHook postConfigure
    '';
  })) pkgs.linuxPackages.kernel;

  kernelPackages = pkgs.linuxPackagesFor kernel;
in {
  boot.kernelPackages = kernelPackages;
}
```

**Pull request closed without merging**